.github/actions/update-extension-provider-images/update-images.py

```diff
@@ -0,0 +1,626 @@
+    """
+    Check if a specific image tag exists in the registry by trying to fetch its manifest.
+
+    Args:
+        oci_client (oci_client.Client): An initialized OCI client.
+        repository (str): The image repository.
+        tag (str): The image tag.
+
+    Returns:
+        bool: True if the image exists, False otherwise.
+    """
+    image_ref = f"{repository}:{tag}"
+    retries = 3
+    delay = 2  # seconds
+
+    for i in range(retries):
+        try:
+            oci_client.manifest(image_reference=image_ref, accept="*/*")
+            return True
+        except OciImageNotFoundException:
+            print(f"  ✗ Manifest for {image_ref} not found.", file=sys.stderr)
+            return False
+        except Exception as e:
+            # Check if the error is a rate-limiting error
+            if "429" in str(e) and "Too Many Requests" in str(e):
+                if i < retries - 1:
+                    print(
+                        f"  ! Rate limited on {image_ref}. Retrying in {delay}s...", file=sys.stderr
+                    )
+                    time.sleep(delay)
+                    delay *= 2  # Exponential backoff
+                    continue
+
+            # For all other errors, or if retries are exhausted
+            print(f"  ✗ Could not find manifest for {image_ref}", file=sys.stderr)
+            print(f"    Error: {e}", file=sys.stderr)
+            return False
+
+    return False
+
+
+# --- Core Logic Functions ---
+def find_newer_versions(
+    oci_client: oci_client.Client,
+    repository: str,
+    current_tags: List[str],
+    available_tags: List[str],
+) -> Dict[str, List[str]]:
+    """
+    Compare current tags with available tags to find newer patch and minor/major versions.
+
+    Args:
+        oci_client (oci_client.Client): An initialized OCI client.
+        repository (str): The image repository.
+        current_tags (List[str]): A list of current tag strings from images.yaml.
+        available_tags (List[str]): A list of all available tag strings from the registry.
+
+    Returns:
+        Dict[str, List[str]]: A dictionary with 'patch', 'minor', and 'missing' keys
+        containing lists of new tags as strings.
+    """
+    available_versions_map = {}
+    for tag in available_tags:
+        try:
+            ver = parse_to_semver(tag)
+            if is_final(ver):
+                available_versions_map[ver] = tag
+        except ValueError:
+            continue
+
+    available_versions = sorted(available_versions_map.keys())
+
+    current_versions = []
+    for tag in current_tags:
+        try:
+            current_versions.append(parse_to_semver(tag))
+        except ValueError:
+            continue
+    current_versions.sort()
+
+    if not current_versions:
+        existing_minor_tags = []
+        missing_minor_tags = []
+        for ver in available_versions:
+            tag = available_versions_map[ver]
+            if image_exists(oci_client, repository, tag):
```

Comment:
I do not quite get why this check is required. To my understanding, the tags returned from OCI-Repository are guaranteed to exist. The only corner-case might be that a tag might have been removed immediately after you requested list-tags. Which is a corner-case that (IMO) does not really suggest (to me at least) this should be double-checked (also, even then, the tag could still be removed after your check..)